Libs/MRML/Core/vtkMRMLModelNode.cxx

```diff
@@ -141,64 +173,130 @@ void vtkMRMLModelNode::SetAndObservePolyData(vtkPolyData *polyData)
+    {
+    return polydata;
+    }
+  else
+    {
+    this->GeometryFilter->Update();
+    return this->GeometryFilter->GetOutput();
+    }
+}
+
+//---------------------------------------------------------------------------
+vtkUnstructuredGrid* vtkMRMLModelNode::GetUnstructuredGrid()
+{
+  return vtkUnstructuredGrid::SafeDownCast(this->GetMesh());
 }
 
 //---------------------------------------------------------------------------
 void vtkMRMLModelNode
-::SetPolyDataConnection(vtkAlgorithmOutput *newPolyDataConnection)
+::SetMeshConnection(vtkAlgorithmOutput *newMeshConnection)
 {
-  if (newPolyDataConnection == this->PolyDataConnection)
+  if (newMeshConnection == this->MeshConnection)
     {
     return;
     }
 
-  vtkAlgorithm* oldPolyDataAlgorithm = this->PolyDataConnection ?
-    this->PolyDataConnection->GetProducer() : 0;
+  vtkAlgorithm* oldMeshAlgorithm = this->MeshConnection ?
+    this->MeshConnection->GetProducer() : 0;
 
-  this->PolyDataConnection = newPolyDataConnection;
+  this->MeshConnection = newMeshConnection;
 
-  vtkAlgorithm* polyDataAlgorithm = this->PolyDataConnection ?
-    this->PolyDataConnection->GetProducer() : 0;
-  if (polyDataAlgorithm != NULL)
+  vtkAlgorithm* newMeshAlgorithm = this->MeshConnection ?
+    this->MeshConnection->GetProducer() : 0;
+  if (newMeshAlgorithm != NULL)
     {
     vtkEventBroker::GetInstance()->AddObservation(
-      polyDataAlgorithm, vtkCommand::ModifiedEvent, this, this->MRMLCallbackCommand );
-    polyDataAlgorithm->Register(this);
+      newMeshAlgorithm, vtkCommand::ModifiedEvent, this, this->MRMLCallbackCommand );
+    newMeshAlgorithm->Register(this);
     }
 
-  this->SetPolyDataToDisplayNodes();
+  this->SetMeshToDisplayNodes();
 
-  if (oldPolyDataAlgorithm != NULL)
+  if (oldMeshAlgorithm != NULL)
     {
     vtkEventBroker::GetInstance()->RemoveObservations (
-      oldPolyDataAlgorithm, vtkCommand::ModifiedEvent, this, this->MRMLCallbackCommand );
-    oldPolyDataAlgorithm->UnRegister(this);
+      oldMeshAlgorithm, vtkCommand::ModifiedEvent, this, this->MRMLCallbackCommand );
+    oldMeshAlgorithm->UnRegister(this);
     }
 
   this->StorableModifiedTime.Modified();
   this->Modified();
-  this->InvokeEvent( vtkMRMLModelNode::PolyDataModifiedEvent , this);
+  this->InvokeEvent( vtkMRMLModelNode::MeshModifiedEvent , this);
+}
+
+//---------------------------------------------------------------------------
+void vtkMRMLModelNode
+::SetPolyDataConnection(vtkAlgorithmOutput *newPolyDataConnection)
+{
+  this->MeshType = vtkMRMLModelNode::PolyDataMeshType;
+  this->SetMeshConnection(newPolyDataConnection);
+  this->GeometryFilter->SetInputConnection(0);
+}
+
+//---------------------------------------------------------------------------
+void vtkMRMLModelNode
+::SetUnstructuredGridConnection(vtkAlgorithmOutput *newUnstructuredGridConnection)
+{
+  this->MeshType = vtkMRMLModelNode::UnstructuredGridMeshType;
```

Comment:
For example, it might be needed when somebody tries to retrieve the mesh type before the mesh is read from file. It's a short time, so most likely nothing tries to access mesh type during that short transient. Just try to think about it and test saving/loading thoroughly and see if you need to do something.